librocksdb_sys/crocksdb/c.cc

```diff
@@ -5853,4 +5857,76 @@ void ctitandb_delete_files_in_ranges_cf(
   SaveError(errptr, static_cast<TitanDB*>(db->rep)->DeleteFilesInRanges(
                         cf->rep, &ranges[0], num_ranges, include_end));
 }
+
+/* RocksDB Cloud */
+#ifdef USE_CLOUD
+struct cloud_envoptions_t {
+  CloudEnvOptions rep;
+};
+
+crocksdb_env_t* cloud_env_create() {
+  // Retrieve path to local dir where db is stored,
+  // name of storage bucket and region of cloud store
+  // TODO: Set all from env variables?
+  std::string kDBPath = "/tmp/rocksdb_cloud_durable";
+  std::string kBucketSuffix = "cloud.durable.example.";
+  std::string kRegion = getenv("AWS_REGION");
+
+  // cloud environment config options here
+  CloudEnvOptions cloud_env_options;
+
+  // Store a reference to a cloud env. A new cloud env object should be
+  // associated
+  // with every new cloud-db.
+  std::unique_ptr<CloudEnv> cloud_env;
+
+  // Retrieve aws access keys from env
+  char* keyid = getenv("AWS_ACCESS_KEY_ID");
+  char* secret = getenv("AWS_SECRET_ACCESS_KEY");
+  if (keyid == nullptr || secret == nullptr) {
+    fprintf(
+        stderr,
+        "Please set env variables "
+        "AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY with cloud credentials");
+    return NULL;
+  }
+  cloud_env_options.credentials.access_key_id.assign(keyid);
+  cloud_env_options.credentials.secret_key.assign(secret);
+
+  // Append the user name to the bucket name in an attempt to make it
+  // globally unique. S3 bucket-names need to be globally unique.
+  // If you want to rerun this example, then unique user-name suffix here.
+  char* user = getenv("USER");
+  kBucketSuffix.append(user);
+
+  // create a bucket name for debugging purposes
+  const std::string bucketName = "tikv." + kBucketSuffix;
+
+  // Create a new AWS cloud env Status
+  CloudEnv* cenv;
+  Status s = CloudEnv::NewAwsEnv(Env::Default(), kBucketSuffix, kDBPath,
+                                 kRegion, kBucketSuffix, kDBPath, kRegion,
+                                 cloud_env_options, nullptr, &cenv);
+  if (!s.ok()) {
+    fprintf(stderr, "Unable to create cloud env in bucket %s. %s\n",
```

Comment:
instead of printing error message, `aws_env_create()` should accept an extra `char** errptr` that pass back the error message. Look at `crocksdb_open` for example.

Reply:
What about the interface at Env::new_aws_env()?